linux/bindings/src/thread.rs

```diff
@@ -0,0 +1,271 @@
+    ///
+    /// Add a prefix "krfl-" and ensure that total length should not be greater
+    /// than `TASK_COMM_LEN`, defined in `include/linux/sched.h`.
+    fn format_name(name: &str) -> [u8; crate::TASK_COMM_LEN as _] {
+        const PREFIX: &str = "krfl-";
+        let pre_len = PREFIX.as_bytes().len();
+        let mut name_buf = [0u8; crate::TASK_COMM_LEN as _];
+        name_buf[..pre_len].copy_from_slice(PREFIX.as_bytes());
+
+        // Copy the user specified `name` to buffer, may be truncated.
+        let avail_len = name_buf.len() - pre_len - 1; // The last byte should always be `\0`.
+        let copy_len = name.as_bytes().len().min(avail_len);
+        name_buf[pre_len..pre_len + copy_len].copy_from_slice(&name.as_bytes()[..copy_len]);
+        name_buf
+    }
+
+    /// Creates an instance with a `name`.
+    ///
+    /// It will run the `func` closure when it's started by `wake_up`.
+    pub fn create<F, T>(name: &str, func: F) -> Result<NonNull<crate::task_struct>>
+    where
+        F: FnOnce() -> T,
+        F: Send + 'static,
+        T: Send + 'static,
+    {
+        // Constructs a boxed `ThreadItem`, which will not be drop until the kthread stops.
+        let item_ptr = Box::into_raw(ThreadItem::try_new(func)?);
+        // SAFETY: no safety requirements on this FFI call.
+        let task_ptr = unsafe {
+            crate::kthread_create_on_node(
+                Some(Self::threadfn::<F, T>),
+                item_ptr as _,
+                crate::NUMA_NO_NODE,
+                Self::format_name(name).as_ptr() as _,
+            )
+        };
+        // SAFETY: FFI call.
+        if unsafe { kernel::bindings::IS_ERR(task_ptr as _) } {
+            // SAFETY: `item_ptr` is initialized above, and it should be dropped
+            // here due to the kthread creation failure.
+            unsafe { Box::from_raw(item_ptr) };
+            return Err(ENOMEM);
+        }
+        // SAFETY: it is checked above that `task_ptr` is valid.
+        Ok(unsafe { NonNull::new_unchecked(task_ptr) })
+    }
+
+    unsafe extern "C" fn threadfn<F, T>(data: *mut core::ffi::c_void) -> core::ffi::c_int
+    where
+        F: FnOnce() -> T,
+        F: Send,
+        T: Send,
+    {
+        // Execute the user closure.
+        let mut item = Box::from_raw(data as *mut ThreadItem<T>);
+        item.execute();
+        // Do not drop the item, until the kthread stops.
+        Box::into_raw(item);
+
+        return 0;
+    }
+
+    /// Returns true if the user specified `func` has been finished.
+    pub fn is_finished<T>(&self) -> bool {
+        // SAFETY: the `self.0` is a valid `task_struct`, so the `kthread_data`
+        // is also valid.
+        unsafe {
+            let item_ptr = crate::kthread_data(self.0.get()) as *mut ThreadItem<T>;
+            (*item_ptr).is_finished.load(Ordering::Acquire)
+        }
+    }
+
+    /// Blocks unless or until the current thread's token is made available.
+    pub fn park() {
```

Comment:
The `park` and `unpark` methods are useful in general. But I don't if they are useful to our case. After all, Rust std's `Thread` does not have these methods. So our OS-independent code cannot depend on the existence of these `park` and `unpark` methods.

Same concern applies to `yield_now`, `sleep` and `wake_up`.